### ISSUE TITLE
Composer docs/bugs (clarification)

### ISSUE THREAD

**Issue by till** [cloudfoundry/php-buildpack#309]
(I'm using buildpack 4.3.77)

I'm trying to overcome a couple challenges deploying a PHP application to a CF PaaS (PWS).

I find your use of composer and related docs a little confusing:

_By default, the PHP buildpack uses the composer.json and composer.lock files that reside inside the root directory, or in the directory specified as WEBDIR in your options.json. If you have composer files inside your app, but not in the default directories, use a COMPOSER_PATH environment variable for your app to specify this custom location, relative to the app root directory. Note that the composer.json and composer.lock files must be in the same directory._

`COMPOSER_PATH` is not mentioned in the table before. It would be great to get more detail if this is a leftover from another time or e.g. if this a separate var to point the buildpack to, to store `composer.json` (and `.lock`).

---

Also, is there history available what this means: "if you have composer files inside your app".

I am confused by this statement, because — where else would they be? Composer keeps dependencies local to the application (usually). So this means, `vendor/` should be within the application, not somewhere else.

---

What is the reason for default `lib/vendor/` with this buildpack? Most code seems to expect `vendor/`? Most projects/frameworks follow a pattern where:

```
lib/ (sometimes called src/)
public/ (or web/ or www/)
vendor/
```

I realise there are (still) exceptions to this and people still put all their code in a "public" directory, but working around it like this, defeats the purpose of the buildpack to e.g. run a demo application.

---

I tried to use `COMPOSER_INSTALL_OPTIONS`:

```
{
    "COMPOSER_INSTALL_OPTIONS": "--no-interaction --no-progress",
    "COMPOSER_VENDOR_DIR": "vendor",
    "WEBDIR": "public"
}
```

This led to:

```
Invalid argument - - n o - i n t e r a c t i o n - - n o - p r o g r e s s. Use "composer require - - n o - i n t e r a c t i o n - - n o - p r o g r e s s" instead to add packages to your composer.json.
```

(**Edit**: I guess this should be a list.)

---

Maybe not in scope of this buildpack — is there a validator for everything in `.bp-config/options.json`)? E.g. pre-push, so I don't have to dig around in the buildpack's code to figure this out?

---

Happy to send a PR with doc fixes or discuss further.

**Comment by dmikusa-pivotal** [comment]
>COMPOSER_PATH is not mentioned in the table before. It would be great to get more detail if this is a leftover from another time or e.g. if this a separate var to point the buildpack to, to store composer.json (and .lock).

This should be accurate & functional (I think docs could use a PR here). Let me try to explain, as it's a little confusing because we look for these files a couple of different places by default (trying to accommodate the variations that exist in the way people structure PHP projects).

The buildpack looks at the root of what you push, this is called the app root in buildpack terminology, but for PHP apps it's more like the project root. The next place it looks is in what we call the WEBDIR, which is just the public/htdocs/web or whatever you want to call it folder (WEBDIR is configurable again to accommodate the variation in project structures).

If it exists elsewhere you can put a path relative to the app root in COMPOSER_PATH and that will tell the buildpack where to find those file. I don't think this option is used much because the places we look by default catch most of the places people put these files. In fact, if we're missing a commonly used location let us know and we can add that to the default search list. We don't really want people to need to use this option, it's there as an escape hatch.

>What is the reason for default lib/vendor/ with this buildpack? Most code seems to expect vendor/? Most projects/frameworks follow a pattern where:

`lib/` (the config option LIBDIR, so you can change the name to `src/` or whatever you want) is on the include_path by default. Putting vendor here means you can `include('vendor/blah`)` and things just work. You don't need to worry about the relative paths. Relative paths can be a little tricky as the buildpack will in some cases attempt to move files around to take an unstructured project and make it into a structured project that it can deploy. This behavior has been confusing/unwanted and is being removed in the Cloud Native PHP buildpack (basically v5 of the PHP buildpack).

You can still use relative paths if you want. You just need to make sure a WEBDIR exists, as this will bypass the behavior where the buildpack moves files. Then set `COMPOSER_VENDOR_DIR` to `vendor`, which will make it app_root/vendor which matches the structure you listed.

Long story short, the buildpack isn't trying to force one way or another, just be flexible to support the different ways people want to structure their projects.

>(Edit: I guess this should be a list.)

Yes, you need to make it a list.

Ex: https://github.com/cloudfoundry/php-buildpack/blob/master/extensions/composer/extension.py#L179

>Maybe not in scope of this buildpack — is there a validator for everything in .bp-config/options.json)? E.g. pre-push, so I don't have to dig around in the buildpack's code to figure this out?

No, doc PRs with better examples would be helpful. Especially a fresh set of eyes, to spot the areas which are confusing/ambiguous. 

Happy to answer more questions if you have them.

**Comment by till** [comment]
> Long story short, the buildpack isn't trying to force one way or another, just be flexible to support the different ways people want to structure their projects.

The Buildpack does the opposite right now though. It's not turn-key, as it requires configuration for the majority of all PHP frameworks and demo applications. Best practice is not to rely on `include_path` anymore as composer encourages local dependencies (scoped within the app) and provides optimised autoloaders which...

 1. Take care of finding files (by using mostly absolute paths, unless you tell it not to)
 2. Provide optimisations for classloading with for OpCache (or formerly APC) and general disk I/O.

Examples (which expect `$APPROOT/vendor`):

 - https://github.com/symfony/demo
 - http://www.slimframework.com
 - https://github.com/Laravel-Backpack/Demo

Search for more on Github — they all have `(/)vendor` in `.gitignore`.

If you follow a pattern where you have scripts in a `bin/` folder, then most of these don't work out of the box, as they `../vendor/autoload.php`. That's why most Symfony examples for Cloud Foundry disable/customize `post-install-cmd`. Not because the scripts don't work, but because their invocation is broken by default (and I just happen to spend more time on trying to figure it out). 

It's great that the bp allows flexibility (even if I don't see any merit), but this is using the opposite of what you should be the standard by introducing this `lib/vendor` thing.

**Comment by dmikusa-pivotal** [comment]
The buildpack has to support the lowest common denominator here, we can't rely on people to be using frameworks or even Composer. I would say that now, it's probably skewed a little towards making it easy for that lowest common denominator app, as opposed to any one framework. That means for some frameworks you just need to turn a few knobs and adjust some config to make it work.

I'll be perfectly honest, I don't see that changing with the current incarnation of the buildpack. There's a lot of history & users which are set doing things in the current manner. If we make changes to the default behavior that risks breaking a lot of other users apps. It certainly couldn't be done in a maintenance release.

Where your feedback is invaluable, thank you for it, is with the next generation of the buildpack, which will follow the [cloud native buildpack spec](https://github.com/buildpack/spec). The spec provides greater flexibility for buildpack authors, so we can do more, it also provides a chance to start fresh and introduce potentially breaking changes.

The CNB's are decomposed, so it's a little tricky to get started, but here are the current projects. They get run in this order, one cycle for detect and one for build.

https://github.com/cloudfoundry/php-cnb/ (provides PHP binaries)
https://github.com/cloudfoundry/php-composer-cnb (provides support for Composer)
https://github.com/cloudfoundry/httpd-cnb (optional, but provides HTTPD binaries)
https://github.com/cloudfoundry/php-web-cnb (generates config for PHP, PHP-FPM & web server of choice)

It's early days, but there's enough functionality there to run some demos. I would appreciate if you add some issues on those projects with questions and suggestions for how to they function, patterns we should follow, etc. Thanks!

**Comment by till** [comment]
@dmikusa-pivotal Can I just use any of these buildpacks with Cloudfoundry (PWS)? Or is this a WIP?

**Comment by dmikusa-pivotal** [comment]
There would be an extra step, but it should work. You would need to run the CNB's using `pack`. That tool produces an image, which you could push to Docker hub or elsewhere. You could then use `cf push -o` to run the image on PWS.

https://github.com/buildpack/pack

**Comment by till** [comment]
@dmikusa-pivotal So I "pack" my app, add to Dockerhub and then change my manifest to use the docker image? I'll see if I can try that on the weekend.

**Comment by dfreilich** [comment]
@till hope it worked! As @dmikusa-pivotal said, thanks for your questions and feedback, it's invaluable. I'll close this for now, but I hope you have time to look through some of the next generation, and give us some suggestions for improvements.

**Comment by till** [comment]
Sorry, but I haven't tried. Too many steps and new tools right now.

I was looking to deploy the symfony demo app and it took a bit too long already (reasons outlined in this ticket). Feel free to ping when a new buildpack is on the horizon/usable. I'll def. give it a go and provide feedback.
